계산기 [STEP1] Morgan

devjoon opened the pull request:
@juun97 
안녕하세요 릴라 이번에 계산기 프로젝트에 참여하게 된 Morgan 입니다!
부족한점이 많아 최대한 열심히 피드백 수용하며 배워나가겠습니다
앞으로 잘 부탁드립니다! 🙇‍♂️

### 고민했던 점
- 큐를 구현해본적이 없어서 제가 구현한 방식이 맞는건지 고민이 되었습니다 

- 요구사항중 "CalculatorItemQueue에서 다루는 요소는 CalculateItem 프로토콜을 준수합니다" 가
어떤 뜻으로 쓰인건지 감이 잘 잡히지 않았습니다 CalculateItem을 CalculatorItemQueue에서 채택하는건지 고민을 하였습니다.

- 선택사항중 List로 직접 구현하는 파트가 있는데 지금은 큐로 구현한 부분도 확실하게 완성이 되었다는 느낌이 들지 않아 피드백을 통해 수정이 필요하다고 생각 하였고 또한 리스트로의 작성에 대해 고민(조언을 얻고싶었던점에 작성하였습니다)이 있었는데 이 고민을 해결하고 리스트로 고치는건 어떨지 생각하였습니다.
---

### 조언을 얻고 싶었던 점
- elements를 private으로 선언시 테스트 케이스에서 접근이 안되는것을 확인하였습니다 그러면 이경우 테스트 케이스를 실행할때
private을 제거하고 테스트 후 다시 private을 써 주어야 하는것인지 고민이 되었습니다 아니면 private 상태에서도 테스트 케이스를 실행
할 수 있는 방법이 있을까요?  

- 커밋내역에 대해서 명확한 기준이 잡히지 않아 여러 방면으로 시도를 해보았습니다
한줄에 모든 내용을 다 적는 방식, 추상적으로 제목을 작성하고 내용에 상세하게 적는방식 등등.
이 경우 어떤 방식이 좀더 좋은 방식일지 궁금합니다.

- (고민했던점 3) 이전 프로젝트 까지는 어떻게 작성할지 고민하고 검색을 통해 문법에 대해서 파악을 하고 적용했었더라면 이번 코드의 경우 선택사항중 리스트로의 구현과 같은 부분이 있었습니다. 근데 이런 것들은 제가 생각한다고 쉽게 짤수 있는것이 아니라 대부분의 작성이 끝나있는 예제코드들을 참고를 하며 짜야 한다 느껴져서 그렇게 짠다고 하면 이게 과연 내가 생각하고 내가 만든건가? 에 대한 고민이 깊어졌습니다.
지금은 큐에대한 기능이 무엇이 있는지 파악하고 이걸 코드로 어떻게 짜면 되겠다 고민하여 제작 하였습니다 하지만 지금 리스트와 같은 다른 방식을 구현한다고 하면 어떻게 해야할지 감이 잘 안잡히는데 이런 부분에 대해서는 다른 분들이 작성한 예제 코드를 통해 전체적으로 이해하고 작성해보는것이 맞을까요?
---

devjoon committed:
feat: 단위테스트를 위한 타깃 추가

devjoon committed:
feat: 기본적인 큐 구현 테스트 케이스 구현

devjoon committed:
feat: TDD방식으로 작업방식 변경 peek테스트 케이스 구현

devjoon committed:
feat: 메소드 구현, 테스트 성공

peek 메소드 구현, peek 정수출력 테스트 성공

devjoon committed:
feat: 테스트 케이스 추가

숫자가 아닌 사칙연산 기호를 enqueue로 실행시 잘 들어가는지에 대한 테스트 케이스 추가

devjoon committed:
refactor: 테스트케이스 수정

devjoon committed:
refactor: 테스트 케이스에 맞춰 dequeue 수정 isEmpty메서드 불필요한 부분 제거

devjoon committed:
refactor: 컨벤션 수정

devjoon committed:
refactor: 컨벤션 수정

juun97 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,37 @@
+//
+import Foundation
+
+protocol CalculateItem {
+    
+}
+
+struct CalculatorItemQueue<T>: CalculateItem {
+    var elements: [T] = []
```
Queue를 struct 로 구현하신 이유가 있을까요?

CalculateItem 이라는 프로토콜을 준수한 Item 들만을 요소로 들게하려면 어떻게 해야할까요?

devjoon replied:
- class가 아닌 struct로 구현한 이유는 지금은 상속의 개념이 필요하지 않을것 같다 생각하여 권장되는 방식인 struct로 시작하게 되었습니다.

devjoon replied:
- 큐의 요소에 CalculateItem protocol을 채택해주었으며  extension을 통해 Int자료형이 CalculateItem protocol을 채택하도록 하였습니다.

juun97 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,37 @@
+        elements.append(element)
+    }
+    
+    mutating func dequeue() {
+        if !elements.isEmpty {
+            elements.removeFirst()
+        }
+    }
```
removeFirst 의 시간복잡도에 대해서 알고 계시나요?

dequeue 가 불완전하게 만들어진 것 같습니다. Queue 에 대해서 좀더 알아보시겠어요?

devjoon replied:
- removeFirst의 경우 시간복잡도는 O(n)이 걸리게 됩니다 이유는 가장 앞의 요소를 제거했을경우 뒤의 요소들을 한칸씩 당겨야하기 때문입니다. 이 부분은 연결리스트로 변경하여 시간복잡도를 O(1)로 단축하였습니다.

- dequeue에 대한 이해가 부족했었던것 같습니다 . 큐에 대해서 다시 공부하게 되었고 연결리스트를 통해 dequeue 실행시 가장 처음들어온 요소의 값을 반환하며 제거하는 코드로 구현하였으며 요소가 없을경우 nil을 반환하도록 구현하였습니다.

juun97 commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,78 @@
+        sut = nil
+    }
+    
+    func test_enqueue실행시_3_2_1를_전달했을때_elements에_잘들어가는지() {
+        sut.enqueue(input: 3)
+        sut.enqueue(input: 2)
+        sut.enqueue(input: 1)
+        
+        XCTAssertEqual(sut.elements, [3,2,1])
+    }
```
test 상황에서

1. give : 어떤 값이 주어 졌을 때
2. when: 어떤 상황이 일어 났을 때
3. then : 결과가 어떻게 나오는가

주석을 사용해 3가지 상황을 나눠서 작성해보시면 테스트 코드의 파악이 더 수월해 집니다.

devjoon replied:
- 처음에는 수행시 제약사항중 "코드에 주석을 남기지 않습니다"  라는 내용이 있어 주석을 사용하지 않고 작업하였습니다. 피드백 이후 리뷰어님 말씀대로 given, when, then 으로 주석을 달고 작업한 결과 좀 더 직관적으로 파악이 가능했습니다. 감사합니다

juun97 commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,78 @@
+        XCTAssertEqual(sut.elements, [3,2,1])
+    }
+    
+    func test_elements가_비어있을시_dequeue실행이_가능한지() {
+        sut.dequeue()
+        
+        XCTAssertEqual(sut.elements, [])
+    }
```
현재 테스트 코드는 정상적으로 테스트가 이루어지고 있지 않습니다. 
해당 테스트 케이스가 실행되면 자동으로 setUpWithError 메서드가 실행되어 sut 을 초기화 해줄텐데 초기화된 sut 과 임의의 배열을 비교 하는것이 현재 테스트 케이스의 이름과 맞는지는 잘 모르겠습니다.

devjoon replied:
- 해당 테스트케이스는 연결리스트로 변경하면서 노드가_없을때_dequeue실행시_nil을_반환하는지() 테스트 케이스로 변경되었습니다 감사합니다.

juun97 commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,78 @@
+        XCTAssertEqual(sut.elements, [])
+    }
+    
+    func test_enqueue로_3_2_1넣고_dequeue_한_번_실행시_elements가_앞에서_하나줄어드는지() {
+        sut.enqueue(input: 3)
+        sut.enqueue(input: 2)
+        sut.enqueue(input: 1)
+        sut.dequeue()
+        
+        XCTAssertEqual(sut.elements, [2,1])
+    }
```
네이밍이 바뀌거나 내부 테스트 로직이 바뀌어야 할 것 같습니다.
네이밍은 elements의 count 를 얘기하고 있지만 실제 로직을 elements 그 자체를 비교하고 있네요.

devjoon replied:
- 로직에 적절하지 못한 네이밍에 대한 오류 확인하였습니다 감사합니다.
- 해당 테스트케이스는 연결리스트로 변경하면서 제거 되었습니다.

juun97 commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,78 @@
+        XCTAssertFalse(sut.isEmpty())
+    }
+    
+    func test_enqueue로_3_2_1넣고_peek_실행시_3이_출력되는지() {
+        sut.enqueue(input: 3)
+        sut.enqueue(input: 2)
+        sut.enqueue(input: 1)
+        
+        let result = sut.peek()
+        
+        XCTAssertEqual(result, 3)
+    }
```
성공코드가 있다면 반대로 실패코드도 테스트를 해주는것이 좋습니다.

devjoon replied:
- 연결리스트로 변경하며 test_노드가_존재하지_않을때_peek실행시_nil값이_출력되는지() 테스트케이스를 추가하였습니다 그 외에도 반대가능성을 생각하여 추가적인 테스트케이스를 작성하였습니다 감사합니다.

juun97 reviewed (commented):
안녕하세요 모건 계산기 STEP 1 고생하셨습니다 😃
큐나 테스트 코드나 전체적으로 깔끔하게 작성해주셨습니다만 아직 큐에 대해 조금 더 알아보셔야 할 것 같습니다. (특히 dequeue)

---

## 고민했던 점 
> 큐를 구현해본적이 없어서 제가 구현한 방식이 맞는건지 고민이 되었습니다

코멘트에서 답변 드리겠습니다.

> 요구사항중 "CalculatorItemQueue에서 다루는 요소는 CalculateItem 프로토콜을 준수합니다" 가
어떤 뜻으로 쓰인건지 감이 잘 잡히지 않았습니다 CalculateItem을 CalculatorItemQueue에서 채택하는건지 고민을 하였습니다.

자세히 읽어 보시면 요소가 프로토콜을 준수하는 것입니다. 그렇다면 Queue 가 아닌 Queue에 들어갈 Element 들이 준수를 하고 있어야 합니다.

> 선택사항중 List로 직접 구현하는 파트가 있는데 지금은 큐로 구현한 부분도 확실하게 완성이 되었다는 느낌이 들지 않아 피드백을 통해 수정이 필요하다고 생각 하였고 또한 리스트로의 작성에 대해 고민(조언을 얻고싶었던점에 작성하였습니다)이 있었는데 이 고민을 해결하고 리스트로 고치는건 어떨지 생각하였습니다.

가장 기본이 되는 Array 를 사용한 구현을 숙지해야 그 뒤에 심화된 Linked List 를 이용한 Queue 구현도 가능하다고 생각합니다. 막히는 부분이 있다면 언제든 도움 드릴테니 우선 Array 를 사용한 Queue 를 먼저 숙지하시고 차근차근 시도 해보면 좋을 것 같습니다.

## 조언을 얻고 싶었던 점
> elements를 private으로 선언시 테스트 케이스에서 접근이 안되는것을 확인하였습니다 그러면 이경우 테스트 케이스를 실행할때
private을 제거하고 테스트 후 다시 private을 써 주어야 하는것인지 고민이 되었습니다 아니면 private 상태에서도 테스트 케이스를 실행
할 수 있는 방법이 있을까요?

테스트는 결국 테스트일뿐 입니다. 테스트를 위해서 본코드가 변경이 된다면 그야 말로 주객이 전도 된 것이죠. 

만약 private 한 메서드를 테스트 해야 한다면 해당 메서드를 사용하고있는 공개된 메서드에서 테스트를 하고자 하는 메서드를 간접적으로 확인 하는 식의 방법이 있고, 추후에 배우실테지만 Test Double 이란 개념이 있습니다. 해당 개념을 사용해 테스트 코드에서만 테스트에 용이한 로직으로 바꾸어 사용할 수 도 있습니다.

> 커밋내역에 대해서 명확한 기준이 잡히지 않아 여러 방면으로 시도를 해보았습니다
한줄에 모든 내용을 다 적는 방식, 추상적으로 제목을 작성하고 내용에 상세하게 적는방식 등등.
이 경우 어떤 방식이 좀더 좋은 방식일지 궁금합니다.

커밋은 결국 사람과 사람사이의 소통 방식입니다. 이 말은 기준 또한 다 다르다는 것이죠. 이 부분은 그 때마다 같이 작업하는 사람들과 의견을 나누고 더 적절하다고 생각하는 방식을 따르는게 나을 것 같습니다.

> (고민했던점 3) 이전 프로젝트 까지는 어떻게 작성할지 고민하고 검색을 통해 문법에 대해서 파악을 하고 적용했었더라면 이번 코드의 경우 선택사항중 리스트로의 구현과 같은 부분이 있었습니다. 근데 이런 것들은 제가 생각한다고 쉽게 짤수 있는것이 아니라 대부분의 작성이 끝나있는 예제코드들을 참고를 하며 짜야 한다 느껴져서 그렇게 짠다고 하면 이게 과연 내가 생각하고 내가 만든건가? 에 대한 고민이 깊어졌습니다.
지금은 큐에대한 기능이 무엇이 있는지 파악하고 이걸 코드로 어떻게 짜면 되겠다 고민하여 제작 하였습니다 하지만 지금 리스트와 같은 다른 방식을 구현한다고 하면 어떻게 해야할지 감이 잘 안잡히는데 이런 부분에 대해서는 다른 분들이 작성한 예제 코드를 통해 전체적으로 이해하고 작성해보는것이 맞을까요?

물론 다른 사람의 코드를 보면서 영감을 얻는 것이 나쁜것만은 아닙니다. 하지만 해당 개념을 알고 코드를 보는것과 모르고 코드를 보는것은 매우 큰 차이가 있습니다. 현재 모건은 Linked List 란 개념에 대해 생소한 것으로 보입니다. 그렇다면 지금 부터 Node 가 무엇인지 이 Node 들을 어떻게 연결하는지 이런 개념들을 학습하시고 처음 부터 천천히 만들어 보다 그 떄도 이해가 안간다면 한 번쯤 도움을 받는 것이 좋을 것 같습니다.

devjoon committed:
refactor: 배열방식의 큐 구현에서 리스트 방식으로 변경시작 , 노드 구현

devjoon committed:
refactor: 기존 enqueue,dequeu,peek 메서드를 연결리스트로 전환, TDD 방식으로 작업 변경

devjoon committed:
feat: isEmpty 메서드를 테스트하는 테스트케이스 추가

devjoon committed:
fix: isEmpty 메서드를 테스트하는 2번째 테스트케이스 추가 후 에러 발생 , 버그 수정

devjoon committed:
feat: clear 메서드를 테스트하는 테스트케이스 추가

devjoon committed:
feat: clear 메서드 구현 , 테스트 성공

devjoon committed:
feat: count 메서드를 테스트 하는 테스트케이스 추가

devjoon committed:
feat: count 메서드 테스트 성공

devjoon committed:
refactor: 테스트케이스 given,when,then처리 , 테스트케이스 수정

test_enqueue로_3을_넣어주고_dequeue로_제거후_isEmpty실행시_true를_반환하는지() 테스트 케이스 코드중 dequeue값을 받는 부분을 와일드카드패턴을 이용하여 수정

devjoon committed:
feat: 테스트 케이스 추가, 테스트 성공, 테스트케이스 네이밍변경

devjoon committed:
feat:큐의 요소를 CalculateItem 프로토콜을 채택하도록 설정

juun97 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,87 @@
+// CalculatorItemQueue
+import Foundation
+
+protocol CalculateItem {
+    
+}
+
+extension Int: CalculateItem {
+    
+}
```
CalculatorItemQueue 와 상관없는 코드들이 존재하네요.
코드의 양이 적더라도 분리를 습관화 해주세요.

devjoon replied:
- CalculatorItem.swift 파일을 만들어서 따로 분리하였습니다.

juun97 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,87 @@
+
+struct CalculatorItemQueue<T: CalculateItem> {
+    var head: Node<T>?
+    var tail: Node<T>?
```
head와 tail은 Queue 가 알아야할 정보일까요? Queue가 관리하는 List 가 알아야할 정보가 아닐까요? 또한 head와 tail 이 외부에서 접근이 필요한 프로퍼티인가요?

devjoon replied:
- LinkedList 로 분리하여 큐가 접근하는 형태로 변경하였습니다 head 와 tail은 외부 접근을 할 수 없게 private처리 하였습니다

juun97 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,87 @@
+        tail = nil
+    }    
+    
+    func count() -> Int {
+        var count = 0
+        guard !isEmpty() else {
+            return count
+        }
+        count += 1
+        var currentNode = head
+        
+        while currentNode?.next != nil {
+            count += 1
+            currentNode = currentNode?.next
+        }
+        return count
+    }
+    
+}
```
해당 로직은 O(n)의 시간복잡도를 가지게 되겠네요. 어떻게 하면 O(1) 의 시간복잡도를 가질 수 있도록 할 수 있을까요?

devjoon replied:
- CalculatorItemQueue에 nodeCount 변수를 생성하고 enqueue 발생시 +1 dequeue 성공시 -1 로 처리하여 count() 함수는 nodeCount의 값만 전달하는 형태로 만들어 O(1) 로 변경하였습니다.

juun97 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,87 @@
+    
+}
+
+class Node<T> {
+    var next: Node<T>?
+    var element: T
+    
+    init(_ element: T) {
+        next = nil
+        self.element = element
+    }
+}
```
Node가 외부에서 접근이 필요한 객체인가요? 또한 상속이 필요한 객체인가요?

devjoon replied:
- 상속이 불필요하여 final을 이용하여 상속불가 처리 하였습니다
- 값을 가져와야 하는 과정이 있어 외부에서 접근은 가능하되 수정이 불가능하도록 private(set)으로 처리하였습니다

juun97 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,87 @@
+    var head: Node<T>?
+    var tail: Node<T>?
+    
+    mutating func enqueue(_ elemnet: T) {
+        let newNode = Node(elemnet)
+        if head == nil {
+            head = newNode
+            tail = newNode
+        } else {
+            tail?.next = newNode
+            tail = newNode
+        }
+    }
+    
+    mutating func dequeue() -> T? {
+        guard let dequeuedElement = head?.element else {
+            return nil
+        }
+        head = head?.next
+        return dequeuedElement
+    }
```
List 를 따로 구현하지 않았기 때문에 Node 객체의 프로퍼티를 직접 접근해서 수정을 해야만 하네요.
위 접근 방법은 OOP 와 관련된 디미터 법칙을 위반하고 있습니다.

devjoon replied:
- Node에 setNext() 메서드를 포함시켜 외부에서 직접 접근해서 수정하는 방식에서 내부에서 수정하는 방식으로 변경하였습니다.

juun97 reviewed (commented):
Linked List 를 사용한 Queue 구현을 시도하신 점 좋습니다 👍🏻
하지만 아직 접근제어의 습관화가 잘 이루어지지 않는 것 같습니다😢
코멘트 남긴 부분 수정완료되면 머지하도록 하겠습니다 🙆🏻‍♂️

devjoon committed:
refactor: 프로토콜 파일분리, count메서드 방식수정, 접근제어처리, enqueue의 OOP준수를 위한 Node의 setNext메서드 추가

devjoon committed:
refactor: LinkedList 를 분리하여 큐에서 접근하는 형태로 리펙토링
